packages/devtools_app/lib/src/network/http_request_data.dart

```diff
@@ -0,0 +1,246 @@
+  double get timeDiffMs => _timeDiffMs;
+  
+  // This is set from within HttpRequestData.
+  double _timeDiffMs;
+}
+
+/// An abstraction of an HTTP request made through dart:io.
+class HttpRequestData {
+  HttpRequestData._(this._timelineMicrosBase, this._startEvent, this._endEvent);
+
+  /// Build an instance from timeline events.
+  /// 
+  /// `timelineMicrosBase` is the offset used to determine the wall-time of a
+  /// timeline event. `events` is a list of Chrome trace format timeline
+  /// events.
+  factory HttpRequestData.fromTimeline(
+      int timelineMicrosBase, List<Map> events) {
+    Map startEvent;
+    Map endEvent;
+    final instantEvents = <Map>[];
+
+    for (final event in events) {
+      if (_isStartEvent(event)) {
+        assert(startEvent == null);
+        startEvent = event;
+      } else if (_isEndEvent(event)) {
+        assert(endEvent == null);
+        endEvent = event;
+      } else if (_isInstantEvent(event)) {
+        instantEvents.add(event);
+      } else {
+        assert(false, 'Unexpected event type');
+      }
+    }
+    final data = HttpRequestData._(timelineMicrosBase, startEvent, endEvent);
+    data._addInstantEvents(
+        [for (final instant in instantEvents) HttpInstantEvent._(instant)]);
+    return data;
+  }
+
+  /// Merges the information from another [HttpRequestData] into this instance.
+  void merge(HttpRequestData data) {
+    if (data.instantEvents.isNotEmpty) {
+      _addInstantEvents(data.instantEvents);
+    }
+    if (data._endEvent != null) {
+      _endEvent = data._endEvent;
+    }
+  }
+
+  // Timeline event helpers.
+  static bool _isStartEvent(Map event) => event['ph'] == 'b';
+  static bool _isEndEvent(Map event) => event['ph'] == 'e';
+  static bool _isInstantEvent(Map event) => event['ph'] == 'n';
+
+  static List<Cookie> _parseCookies(List cookies) {
+    return [for (final cookie in cookies) Cookie.fromSetCookieValue(cookie)];
+  }
+
+  void _addInstantEvents(List<HttpInstantEvent> events) {
+    _instantEvents.addAll(events);
+
+    // This event is the second half of an outstanding request which will be
+    // merged into a single HttpRequestData elsewhere. We'll calculate the
+    // instant event times then since we'll have _startEvent's timestamp.
+    if (_startEvent == null) {
+      return;
+    }
+    _recalculateInstantEventTimes();
+  }
+
+  void _recalculateInstantEventTimes() {
+    assert(_startEvent != null);
+    int lastTime = _requestTimeMicros;
+    for (final instant in instantEvents) {
+      final instantTime = _getTimelineMicrosecondsSinceEpoch(instant._rawEvent);
+      instant._timeDiffMs = (instantTime - lastTime) / 1000;
+      lastTime = instantTime;
+    }
+  }
+
+  int _getTimelineMicrosecondsSinceEpoch(Map event) {
+    assert(event.containsKey('ts'));
+    return _timelineMicrosBase + event['ts'];
+  }
+
+  /// The duration of the HTTP request, in milliseconds.
+  double get durationMs {
```

Comment:
Use the TimeRange class to store timing information. This is used in the Timeline.